File: app/src/main/java/com/vn/ezlearn/adapter/NavigationAdapter.java
package com.vn.ezlearn.adapter;

import android.animation.ObjectAnimator;
import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.SparseBooleanArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.github.aakira.expandablelayout.ExpandableLayout;
import com.github.aakira.expandablelayout.ExpandableLayoutListenerAdapter;
import com.github.aakira.expandablelayout.ExpandableLinearLayout;
import com.github.aakira.expandablelayout.Utils;
import com.vn.ezlearn.R;
import com.vn.ezlearn.models.ItemMenu;
import com.vn.ezlearn.models.ItemMenuChild;
import com.vn.ezlearn.interfaces.NavigationItemSelected;
import com.vn.ezlearn.widgets.CRecyclerView;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by FRAMGIA\nguyen.duc.manh on 07/07/2017.
 */

public class NavigationAdapter extends BaseRecyclerAdapter<ItemMenu, NavigationAdapter.ViewHolder> {

    private final List<ItemMenu> data;
    private Context context;
    private SparseBooleanArray expandState = new SparseBooleanArray();
    private NavigationItemSelected navigationItemSelected;

    public NavigationAdapter(Context context, List<ItemMenu> list,
                             NavigationItemSelected navigationItemSelected) {
        super(context, list);
        this.data = list;
        this.navigationItemSelected = navigationItemSelected;
        for (int i = 0; i < data.size(); i++) {
            expandState.append(i, false);
        }
    }

    @Override
    public ViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
        this.context = parent.getContext();
        if (viewType == ItemMenu.TYPE_NORMAL) {
            return new ViewHolder(LayoutInflater.from(context)
                    .inflate(R.layout.item_menu_normal, parent, false));
        } else {
            return new ViewHolder(LayoutInflater.from(context)
                    .inflate(R.layout.item_menu, parent, false));
        }

    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position) {
        final ItemMenu item = data.get(position);
        holder.setIsRecyclable(false);
        holder.tvName.setText(item.name);

        if (getItemViewType(position) != ItemMenu.TYPE_NORMAL) {
            NavigationChildAdapter childAdapter = new NavigationChildAdapter(
                    context, new ArrayList<ItemMenuChild>(), navigationItemSelected);
            holder.rvItemMenuChild.setAdapter(childAdapter);
            childAdapter.addAll(item.menuChildList);
            holder.expandableLayout.setInRecyclerView(true);
            holder.expandableLayout.setInterpolator(Utils.createInterpolator(Utils.FAST_OUT_SLOW_IN_INTERPOLATOR));
            holder.expandableLayout.setExpanded(expandState.get(position));
            holder.expandableLayout.setListener(new ExpandableLayoutListenerAdapter() {
                @Override
                public void onPreOpen() {
                    createRotateAnimator(holder.buttonLayout, 0f, 180f).start();
                    expandState.put(position, true);
                }

                @Override
                public void onPreClose() {
                    createRotateAnimator(holder.buttonLayout, 180f, 0f).start();
                    expandState.put(position, false);
                }
            });

            holder.buttonLayout.setRotation(expandState.get(position) ? 180f : 0f);
            holder.buttonLayout.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(final View v) {
                    onClickButton(holder.expandableLayout);
                }
            });
            holder.item.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    onClickButton(holder.expandableLayout);
                }
            });
        } else {
            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    navigationItemSelected.onSelected(item.name, item.id);
                }
            });
        }

    }

    private void onClickButton(final ExpandableLayout expandableLayout) {
        expandableLayout.toggle();
    }

    @Override
    public int getItemViewType(int position) {
        return data.get(position).type;
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public TextView tvName;
        public RelativeLayout buttonLayout;
        public CRecyclerView rvItemMenuChild;
        public ExpandableLinearLayout expandableLayout;
        public RelativeLayout item;

        public ViewHolder(View v) {
            super(v);
            tvName = v.findViewById(R.id.tvName);
            buttonLayout = v.findViewById(R.id.button);
            item = v.findViewById(R.id.item);
            expandableLayout = v.findViewById(R.id.expandableLayout);
            rvItemMenuChild = v.findViewById(R.id.rvItemMenuChild);
        }
    }

    public ObjectAnimator createRotateAnimator(final View target, final float from, final float to) {
        ObjectAnimator animator = ObjectAnimator.ofFloat(target, "rotation", from, to);
        animator.setDuration(300);
        animator.setInterpolator(Utils.createInterpolator(Utils.LINEAR_INTERPOLATOR));
        return animator;
    }
}

File: app/src/main/java/com/vn/ezlearn/adapter/NavigationAdapter.kt
package com.vn.ezlearn.adapter

import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.support.v7.widget.RecyclerView
import android.util.SparseBooleanArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import android.widget.TextView
import com.github.aakira.expandablelayout.ExpandableLayout
import com.github.aakira.expandablelayout.ExpandableLinearLayout
import com.github.aakira.expandablelayout.Utils
import com.vn.ezlearn.R
import com.vn.ezlearn.config.GlobalValue
import com.vn.ezlearn.interfaces.NavigationItemChildSelected
import com.vn.ezlearn.interfaces.NavigationItemSelected
import com.vn.ezlearn.models.Category
import com.vn.ezlearn.widgets.CRecyclerView

/**
 * Created by FRAMGIA\nguyen.duc.manh on 07/07/2017.
 */

class NavigationAdapter(context: Context, list: MutableList<Category>,
                        private val navigationItemSelected: NavigationItemSelected) :
        BaseRecyclerAdapter<Category, NavigationAdapter.ViewHolder>(context, list),
        NavigationItemChildSelected {

    override fun onSelected(name: String, id: String, categoryList: List<Category>?,
                            parentPosition: Int) {
//        CLog.d(AppUtils.getTAG(NavigationAdapter::class.java), "position_menu_parent_old: "
//                + GlobalValue.position_menu_parent_old + " " + parentPosition)
//        if (GlobalValue.position_menu_parent_old != parentPosition) {
//            if (GlobalValue.position_menu_parent_old > 0)
//                list[GlobalValue.position_menu_parent_old].childAdapter!!.clearBackground(true)
//            GlobalValue.position_menu_parent_old = parentPosition
//        }
//        GlobalValue.position_menu_old = parentPosition
//        clearBackground()
        navigationItemSelected.onSelected(name, id, categoryList)
    }

    private val data: List<Category>
    private var context: Context? = null
    private val expandState = SparseBooleanArray()

    init {
        this.data = list
        for (i in data.indices) {
            expandState.append(i, true)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        this.context = parent.context
        return when (viewType) {
            Category.TYPE_NORMAL -> ViewHolder(LayoutInflater.from(context)
                    .inflate(R.layout.item_menu_normal, parent, false))
            Category.TYPE_LINE -> ViewHolder(LayoutInflater.from(context)
                    .inflate(R.layout.item_menu_line, parent, false))
            else -> ViewHolder(LayoutInflater.from(context)
                    .inflate(R.layout.item_menu, parent, false))
        }

    }

    override fun onBindViewHolder(holder: ViewHolder, @SuppressLint("RecyclerView") position: Int) {
        if (getItemViewType(position) != Category.TYPE_LINE) {
            val item = data[position]
            holder.setIsRecyclable(false)
            holder.tvName?.text = item.category_name

            if (getItemViewType(position) != Category.TYPE_NORMAL) {
                val childAdapter = NavigationChildAdapter(
                        mContext, ArrayList(), this, position)
                item.childAdapter = childAdapter
                holder.rvItemMenuChild?.adapter = childAdapter
                childAdapter.addAll(item.children!!)
                holder.expandableLayout?.setInRecyclerView(true)
                holder.expandableLayout?.setInterpolator(Utils.createInterpolator(
                        Utils.FAST_OUT_SLOW_IN_INTERPOLATOR))
                holder.expandableLayout?.isExpanded = expandState.get(position)
//                holder.expandableLayout?.setListener(object : ExpandableLayoutListenerAdapter() {
//                    override fun onPreOpen() {
//                        createRotateAnimator(holder.buttonLayout!!, 0f, 180f).start()
//                        expandState.put(position, true)
//                    }
//
//                    override fun onPreClose() {
//                        createRotateAnimator(holder.buttonLayout!!, 180f, 0f).start()
//                        expandState.put(position, false)
//                    }
//                })

                holder.buttonLayout?.rotation = if (expandState.get(position)) 180f else 0f
//                holder.buttonLayout?.setOnClickListener { onClickButton(holder.expandableLayout!!) }
//                holder.item?.setOnClickListener { onClickButton(holder.expandableLayout!!) }
            } else {
//                if (item.isSelected) {
//                    holder.item!!.setBackgroundResource(R.drawable.bg_item_navigation_press)
//                } else {
//                    holder.item!!.setBackgroundResource(R.drawable.bg_item_navigation)
//                }
                if (item.levelChild!! >= 2) {
                    holder.itemView.setOnClickListener {
//                        setBackgroundPress(position)
                        navigationItemSelected.onSelected(item.category_name, item.category_id,
                                item.children!!)
                    }

                } else {
                    holder.itemView.setOnClickListener {
//                        setBackgroundPress(position)
                        navigationItemSelected.onSelected(item.category_name, item.category_id, null)
                    }
                }

            }

        }

    }

    private fun setBackgroundPress(position: Int) {
        if (position != GlobalValue.position_menu_old) {
            var category = list[GlobalValue.position_menu_old]
            category.isSelected = false
            setData(GlobalValue.position_menu_old, category)
            category = list[position]
            category.isSelected = true
            setData(position, category)
            GlobalValue.position_menu_old = position
            list.indices
                    .filter { list[it].childAdapter != null && list[it].childAdapter!!.itemCount > 0 }
                    .forEach { list[it].childAdapter!!.clearBackground(true) }
            notifyDataSetChanged()
        }
    }


    private fun onClickButton(expandableLayout: ExpandableLayout) {
        expandableLayout.toggle()
    }

    override fun getItemViewType(position: Int): Int = data[position].typeMenu!!

    override fun getItemCount(): Int = data.size

    inner class ViewHolder(v: View) : RecyclerView.ViewHolder(v) {
        var tvName: TextView? = v.findViewById(R.id.tvName)
        var buttonLayout: RelativeLayout? = v.findViewById(R.id.button)
        var rvItemMenuChild: CRecyclerView? = v.findViewById(R.id.rvItemMenuChild)
        var expandableLayout: ExpandableLinearLayout? = v.findViewById(R.id.expandableLayout)
        var item: RelativeLayout? = v.findViewById(R.id.item)

    }

    fun createRotateAnimator(target: View, from: Float, to: Float): ObjectAnimator {
        val animator = ObjectAnimator.ofFloat(target, "rotation", from, to)
        animator.duration = 300
        animator.interpolator = Utils.createInterpolator(Utils.LINEAR_INTERPOLATOR)
        return animator
    }

    private fun clearBackground() {
        var category: Category?
        for (i in list.indices) {
            if (list[i].isSelected) {
                category = list[i]
                category.isSelected = false
                setData(i, category)
            }
        }
        notifyDataSetChanged()

    }
}